File: Data Structures & Algorithms/spiral-matrix/submission-1.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        res=[]
        m,n=len(matrix),len(matrix[0])
        top, bottom, left, right = 0, m-1, 0, n-1

        while top<=bottom and left<=right:
            for a in range(left, right+1):
                res.append(matrix[top][a])
            top+=1

            for b in range(top, bottom+1):
                res.append(matrix[b][right])
            right-=1

            if top>bottom or left>right:
                break

            for c in range(right, left-1, -1):
                res.append(matrix[bottom][c])
            bottom-=1

            for d in range(bottom, top-1, -1):
                res.append(matrix[d][left])
            left+=1

        return res

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    PriorityQueue<Integer> pq;
    public int kthSmallest(TreeNode root, int k) {
        this.pq = new PriorityQueue<>();
        traverse(root);
        if(k>pq.size()) return -1;
        int t=1;
        int res=0;
        while(t<=k) {
            res=pq.poll();
            t++;
        }
        return res;
    }
    
    private void traverse(TreeNode root) {
        if(root!=null) {
            this.pq.add(root.val);
        } else {
            return;
        }

        traverse(root.left);
        traverse(root.right);
    }
}
